text/0470-aurora-serverless-v2.md

```diff
@@ -0,0 +1,173 @@
+# Aurora Serverless v2 support
+
+* **Original Author(s):**: @pahud
+* **Tracking Issue**: #470
+* **API Bar Raiser**: @vinayak-kukreja
+
+Allow users to create Amazon Aurora Serverless v2 instances with the `aws-rds` module.
+
+## Working Backwards
+
+> This section should contain one or more "artifacts from the future", as if the
+> feature was already released and we are publishing its CHANGELOG, README,
+> CONTRIBUTING.md and optionally a PRESS RELEASE. This is the most important
+> section of your RFC. It's a powerful thought exercise which will challenge you
+> to truly think about this feature from a user's point of view.
+>
+> Choose *one or more* of the options below:
+>
+> * **CHANGELOG**: Write the changelog entry for this feature in conventional
+>   form (e.g. `feat(eks): cluster tags`). If this change includes a breaking
+>   change, include a `BREAKING CHANGE` clause with information on how to
+>   migrate. If migration is complicated, refer to a fictional GitHub issue and
+>   add its contents here.
+
+feat(rds): Aurora Serverless v2 support
+
+>
+> * **README**: If this is a new feature, write the README section which
+>   describes this new feature. It should describe the feature and walk users
+>   through usage examples and description of the various options and behavior.
+>
+
+## Aurora Serverless v2 support
+
+Aurora Serverless v2 is an on-demand, autoscaling configuration for Amazon Aurora.
+Aurora Serverless v2 helps to automate the processes of monitoring the workload and
+adjusting the capacity for your databases. Capacity is adjusted automatically based on
+application demand. Read
+[Using Aurora Serverless v2](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.html)
+for more details.
+
+### Create a new cluster
+
+Use `DatabaseClusterV2` to create a new cluster with Aurora Serverless V2 support.
+You may specify the `writer` and `readers` instances with the cluster. The instances
+could be either serverless or privisioned.
+
+```ts
+new rds.DatabaseClusterV2(stack, 'cluster', {
+  engine,
+  vpc,
+  // writer(serverless)
+  writer: { serverless: true },
+  readers: [
+    // reader 1(provisioned)
+    { instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.LARGE) },
+    // reader 2(serverless)
+    { serverless: true },
```

Comment:
This looks odd to me. We have more flexibility provisioning an instance but serverless is just a boolean? Can this experience be modeled to be similar?

Reply:
We need an interface for `writer` and `readers` and they can be either provisioned or serverless and the default is `provisioned`.

For a serverless instance, the minimal required property would be `serverless: true` since the default is provisioned while for a provisioned instanced the default could be just `{}` but you can optionally specify custom `instanceType` which is not allowed in serverless. 

I think this could satisfy both serverless and provisioned here but I am open to different design for this interface.

Reply:
I believe there could be an issue with large clusters with our current API with list of readers and writers.

For instance, imagine a cluster with 50 readers. Instantiating with current API would look quite large and have redundant information as well. 

Also, could you tell me what do you mean by custom instance type is not allowed in serverless? From what I understand, we can have ACU's modified for serverless v2 instances?

Reply:
From my current design, if you are creating a large cluster with 50 or more readers, you need to pass a list of instance options to the `readers` prop. And each element in the list could have different instance types or instance options. Do you expect any other user experience for this scenario?

```ts
new DatabaseClusterV2(scope, id, {
  writer,
  readers: readerList,
}
```

>Also, could you tell me what do you mean by custom instance type is not allowed in serverless? From what I understand, we can have ACU's modified for serverless v2 instances?

In serverlessV2, there is no option to specify individual ACU for individual serverless instances. All serverless instances are sharing the same [ServerlessV2ScalingConfiguration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html#cfn-rds-dbcluster-serverlessv2scalingconfiguration) which is cluster-wise. Behind the scene the serverless instance is literally the [CfnDbInstance](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbinstance.html) which does not allow you to specify any specific ACU for it.